File: ts/src/pipeline/serialize/serialize-context.ts
import type { Annotations } from "../../ir/annotations/annotations.js";
import type { Diagnostic } from "../../ir/diagnostic/diagnostic.js";
import type { Scope } from "../../ir/scope/scope.js";
import type { Language } from "../../language.js";

type SourceMeta = Readonly<{
  path: string;
  language: Language;
}>;

export type SerializeContext = Readonly<{
  rootScope: Scope;
  annotations: Annotations;
  source: SourceMeta;
  diagnostics: readonly Diagnostic[];
  raw: string;
}>;

File: crates/unsnarl-emitter/src/serialize_context.rs
//! `SerializeContext`: the bundle of inputs passed to `IRSerializer::serialize`.
//!
//! Mirrors `SerializeContext` in `ts/src/pipeline/serialize/serialize-context.ts`.

use unsnarl_annotations::Annotations;
use unsnarl_ir::diagnostic::Diagnostic;
use unsnarl_ir::ids::ScopeId;
use unsnarl_ir::language::Language;
use unsnarl_ir::IrArena;

pub struct SerializeSourceMeta {
    pub path: String,
    pub language: Language,
}

pub struct SerializeContext<'a> {
    pub arena: &'a IrArena,
    pub root_scope: ScopeId,
    pub annotations: &'a dyn Annotations,
    pub source: SerializeSourceMeta,
    pub diagnostics: &'a [Diagnostic],
    pub raw: &'a str,
}
